data_structures/src/transaction_factory.rs

```diff
@@ -5,83 +5,123 @@ use crate::{
-    };
+        let utxo_iter = self.sort_by(utxo_strategy);
 
-    for op in utxo_iter.iter() {
-        let ts = own_utxos.get_mut(op).unwrap();
-        let value = all_utxos.get(op).unwrap().value;
-        total += value;
-        if all_utxos.get(op).unwrap().time_lock > timestamp {
-            continue;
-        }
+        for op in utxo_iter.iter() {
+            let value = self.get_value(op).unwrap();
+            total = total
+                .checked_add(value)
+                .ok_or_else(|| TransactionError::OutputValueOverflow)?;
 
-        if timestamp < *ts {
-            continue;
-        }
+            if let Some(time_lock) = self.get_time_lock(op) {
+                if time_lock > timestamp {
+                    continue;
+                }
+            }
 
-        if let Some(block_number_limit) = block_number_limit {
-            // Ignore all outputs created after `block_number_limit`.
-            // Outputs from the genesis block will never be ignored because `block_number_limit`
-            // can't go lower than `0`.
-            if all_utxos.included_in_block_number(op).unwrap() > block_number_limit {
-                continue;
+            if let Some(block_number_limit) = block_number_limit {
+                // Ignore all outputs created after `block_number_limit`.
+                // Outputs from the genesis block will never be ignored because `block_number_limit`
+                // can't go lower than `0`.
+                if let Some(limit) = self.get_included_block_number(op) {
+                    if limit > block_number_limit {
+                        continue;
+                    }
+                } else {
+                    continue;
+                }
             }
-        }
 
-        acc += value;
-        list.push(op.clone());
+            acc += value;
+            list.push(op.clone());
+
+            if acc >= amount {
+                break;
+            }
+        }
 
         if acc >= amount {
-            break;
+            Ok((list, acc))
+        } else {
+            Err(TransactionError::NoMoney {
+                total_balance: total,
+                available_balance: acc,
+                transaction_value: amount,
+            })
         }
     }
 
-    if acc >= amount {
-        // Mark UTXOs as used so we don't double spend
-        for elem in &list {
-            let ts = own_utxos.get_mut(elem).unwrap();
-            // Save the timestamp until it could be spend it
-            *ts = timestamp + tx_pending_timeout;
-        }
-        Ok((list, acc))
-    } else {
-        Err(TransactionError::NoMoney {
-            total_balance: total,
-            available_balance: acc,
-            transaction_value: amount,
-        })
+    /// Generic inputs/outputs builder: can be used to build
+    /// value transfer transactions and data request transactions.
+    #[allow(clippy::too_many_arguments)]
+    fn build_inputs_outputs_inner(
```

Comment:
Now that this is function is included in the OutputsCollection trait, maybe it's time to remove the "inner" from the name?